aedes.js

```diff
@@ -170,7 +170,6 @@ function storeRetained (packet, done) {
 }
 
 function emitPacket (packet, done) {
-  packet.retain = false
```

Comment:
Why? This is used to prevent other broker instances to store this packet again when it's emiitted

Reply:
Unfortunately this cannot be easily removed.

Reply:
We need to keep the retain on the packet until we are able to check if "retain as published" (rap) is true.  That's the whole point of the rap flag.  The bridge protocol enables retain-as-published.  I moved this into the subscription handler.  What are the concerns with moving this into the subscription handler?

Reply:
can you link it to the block of code that replaces this?

Reply:
@mcollina this is what handlePacketSubscription in subscribe.js is doing.  Instead of globally setting it to false for all subscriptions, it sets it to false only if the retain-as-published flag is not set on the subscription.

If clusters go through a different (untested?) path, then clearing the retain flag can happen on that path, or I could set up a different emitter for client subscriptions so that clusters always get the flag set to false while bridges and regular clients respect the rap flag.

Reply:
I think as an alternative we could add a `_stored` property to the packet to prevent others to store it again?

Reply:
I think we should verify if this is still needed, e.g. if the packet gets stored multiple times when using a cluster.

Reply:
This only happens when a user sends a packet using `broker.publish` function. BTW I think that it could be safely removed by double checking the code, the 'worst' thing that could happen is some useless write to the db

Reply:
I had thought of the `_stored` idea, but wasn't sure whether it would cause other issues.  In any case, it becomes the event listener's responsibility to not retain the packet if it's not supposed to.

As I've said, the tests pass, including some new ones I added, so how can I verify this concern?

Reply:
> As I've said, the tests pass, including some new ones I added, so how can I verify this concern?

You should add a test where you do a broker subscribe followed by a broker publish to check if the received message has the retained flag cleared or not

Reply:
I added the test and, indeed, the retain flag was previously set to false and is now true.  I've pushed a change with a proposed solution.  The publish function now takes the `rap` parameter and applies the wrapper function that I had previously placed in `subscription.js`.  This way all callers of `subscribe` have the retain-as-published functionality available to them but default to the standard behaviour of setting `retain` to `false`.